fedimint-api-client/src/api/mod.rs

```diff
@@ -1032,6 +1062,575 @@ where
+    ) -> JsonRpcResult<Value> {
+        let method = match self.module_id {
+            Some(module_id) => ApiMethod::Module(module_id, method.to_string()),
+            None => ApiMethod::Core(method.to_string()),
+        };
+
+        self.connections
+            .request(peer_id, method, params.clone())
+            .await
+            .map_err(|e| JsonRpcClientError::Custom(e.to_string()))
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ReconnectClientConnections {
+    connections: BTreeMap<PeerId, ClientConnection>,
+    #[allow(dead_code)]
+    tg: TaskGroup,
+}
+
+impl ReconnectClientConnections {
+    pub fn new(connector: &DynClientConnector, tg: TaskGroup) -> Self {
+        ReconnectClientConnections {
+            connections: connector
+                .peers()
+                .into_iter()
+                .map(|peer| (peer, ClientConnection::new(peer, connector.clone(), &tg)))
+                .collect(),
+            tg,
+        }
+    }
+
+    async fn request(
+        &self,
+        peer: PeerId,
+        method: ApiMethod,
+        request: ApiRequestErased,
+    ) -> anyhow::Result<Value> {
+        self.connections
+            .get(&peer)
+            .expect("Could not find client connection for peer {peer}")
+            .connection()
+            .context("Peer is disconnected")?
+            .request(method, request)
+            .await
+    }
+}
+
+#[derive(Clone, Debug)]
+struct ClientConnection {
+    receiver: watch::Receiver<Option<DynClientConnection>>,
+}
+
+impl ClientConnection {
+    fn new(
+        peer: PeerId,
+        connector: DynClientConnector,
+        task_group: &TaskGroup,
+    ) -> ClientConnection {
+        let (sender, receiver) = watch::channel(None);
+
+        task_group.spawn_cancellable(
+            format!("peer-api-reconnection-{peer}"),
+            async move {
+                loop {
+                    let connection = retry("reconnect-peer-api", api_networking_backoff(), || {
+                        connector.connect(peer)
+                    })
+                    .await
+                    .expect("Unlimited retries");
+
+                    info!(target: LOG_CLIENT_NET_API, "Connected to peer api");
+
+                    if sender.send(Some(connection.clone())).is_err() {
+                        break;
+                    }
+
+                    connection.await_disconnection().await;
+
+                    info!(target: LOG_CLIENT_NET_API, "Disconnected from peer api");
+
+                    if sender.send(None).is_err() {
+                        break;
+                    }
+                }
+
+                info!(target: LOG_CLIENT_NET_API, "Shutting down peer api reconnection task");
+            }
+            .instrument(info_span!("peer-api-reconnection", ?peer)),
+        );
+
+        ClientConnection { receiver }
+    }
+
+    fn connection(&self) -> Option<DynClientConnection> {
+        self.receiver.borrow().as_ref().cloned()
```

Comment:
If the reconnecting task is in the middle of reconnecting, this will return `None` right away, which will make it look like random rpc attempts are failing sometimes for no other reason than previous connections happened to time out around the same time. Less than ideal, I think. It seems to me that every caller should somehow block here until at least one reconnection attempt on the inner task was made before failing upwards.